raw2ppm/raw2ppm.6
.TH raw2ppm 6 "12 January 2000"
.ds . \fBraw2ppm\fP

.SH NAME
raw2ppm \- convert Doom wad 8-bit raw graphic data to PPM

.SH SYNOPSIS
.B raw2ppm
[\fB\-Iv\fP]
[\fB\-c\fP \fIgame\fP]
[\fB\-h\fP \fIheight\fP]
[\fB\-n\fP \fIcount\fP]
[\fB\-o\fP \fIfile\fP]
[\fB\-s\fP \fIoffset\fP]
[\fB\-w\fP \fIwidth\fP]
.IR file " ..."

.SH DISCLAIMER
This program is furnished on an AS IS basis for the fun of it.
id Software does not support this program.

.SH DESCRIPTION
\*. has two primary uses. The first one is converting flat lumps extracted from
a wad by \fBwadext\fP into graphic files. For this purpose, it has been
obsoleted by \fBwadflat\fP. The second one is extracting certain graphics from
the Doom alpha 0.4 and Doom alpha 0.5 iwads. For \fIthat\fP purpose, \*. is
expected to become obsolete when DeuTex is taught about the Doom alpha graphics
format.
.PP
Even though \*. was originally written for a precise task, it was designed to
be reasonably general, in the hope that it may be useful for other things.
.PP
\*. converts a stream of bytes into an image, on the basis of one byte to one
pixel. The byte values are mapped to RGB colours using the palette specified by
the \fB\-c\fP option (by default, the Doom/Doom II palette).
.PP
The pixels are read from the input file in normal English reading order, i.e.
from left to right and from top to bottom. If the \fB\-I\fP option is
specified, some interleaving is used.
.PP
To choose the geometry of the output image, \*. first determines \fIN\fP,
the number of bytes to read. By default, \fIN\fP is equal to the size of the
input file (this can be changed through \fB\-n\fP and \fB\-s\fP). If you
specify the width (with \fB\-w\fP), \*. automatically sets the height to
\fIN\fP\(di\fIw\fP. Conversely, if you specify the height (with \fB\-h\fP), \*.
sets the width to \fIN\fP\(di\fIh\fP. If you specify neither width nor height,
\*. assumes the width is 64 pixels and sets the height to \fIN\fP\(di\64. If
you specify both, \*. just says amen.
.PP
If \fIw\fP\(mu\fIh\fP is lesser than \fIN\fP, \*. prints a warning and
drops the \fIN\fP\(mi\fIw\fP\(mu\fIh\fP last bytes so that the output image
has no incomplete row. If \fIw\fP\(mu\fIh\fP is greater than \fIN\fP, \*.
prints an error message and skips the file.
.PP
For each input file, the name of the output file is chosen by removing any
"\fB.lmp\fP" extension at the end of the input file name and appending the
"\fB.ppm\fP" extension. For example, for an input file named "foo/bar.lmp", the
output file would be "foo/bar.ppm". If there is only one input file, you can
override the default output file name by using the \fB\-o\fP option.

.SH OPTIONS
All options are global. It is \fInot\fP possible to specify different values
of, for example, the width by interleaving \fB\-w\fP options and \fIfile\fP
arguments. If an option is repeated, the last occurrence wins. \*. parses its
command line with \fBgetopt()\fP and therefore supports all the usual features
(see \fBgetopt\fR(3)). All file names are limited to 256 characters.
.TP
.BI "\-c " game
Translate pixel values to RGB colours using the palette from game \fIgame\fP.
The possible values of \fIgame\fP are \fBdoom\fP, \fBdoom04\fP, \fBdoom04t\fP,
\fBdoom05\fP, \fBdoom05t\fP, \fBheretic\fP, \fBhexen\fP and \fBstrife\fP.
\fBdoom\fP is for Doom and Doom II. \fBdoom04\fP and \fBdoom05\fP are for Doom
alpha 0.4 and Doom alpha 0.5 respectively. \fBdoom04t\fP and \fBdoom05t\fP are
the \fBTITLEPIC\fP palettes (\fBTITLEPAL\fP) of Doom alpha 0.4 and Doom alpha
0.5 respectively. The default is \fBdoom\fP.
.TP
.BI "\-h " height
Height of the output image. \fIheight\fP is a decimal integer unless prefixed
by \fB0x\fP (hexadecimal) or \fB0\fP (octal). \fIheight\fP must be greater
than 0.
.TP
.B \-I
Interleave. Treat the input file as 4 frames of (\fIw\fP\(di4\(mu\fIh\fP)
pixels where the first frames provides the pixels for columns 0, 4, and so on,
the second frame columns 1, 5 and so on, the third frame columns 2, 6 and so
on and the fourth frame columns 3, 7 and so on. Useful to extract certain
graphics from the Doom alpha iwad.
.TP
.BI "\-n " count
Number of bytes to read from each input file. \fIcount\fP is a decimal integer
unless prefixed by \fB0x\fP (hexadecimal) or \fB0\fP (octal). \fIcount\fP must
be positive.
.TP
.BI "\-o " file
Override the default output file name. If you attempt to use this option when
there is more than one input file, \*. refuses to run.
.TP
.BI "\-s " offset
Skip \fIoffset\fP bytes at the beginning of the input file. \fIoffset\fP is a
decimal integer unless prefixed by \fB0x\fP (hexadecimal) or \fB0\fP (octal).
\fIoffset\fP must be positive.
.TP
.B \-v
Verbose mode. Causes \*. to print for each file it processes the string
"%s \-> %s", where the first "%s" is replaced by the name of the input file and
the second one by the name of the output file.
.TP
.BI "\-w " width
Width of the output image. \fIwidth\fP is a decimal integer unless prefixed
by \fB0x\fP (hexadecimal) or \fB0\fP (octal). \fIwidth\fP must be greater
than 0.

.SH SUPPORTED GAMES
Doom, Doom alpha, Doom PR, Ultimate Doom, Doom][, Final Doom, Heretic, Hexen,
Strife.

.SH EXIT VALUE
\fB0\fP	OK.
.br
\fB1\fP	Syntax error.
.br
\fB2\fP	An error occurred on one or more of the files.

.SH SEE ALSO
The Unofficial Doom Specs by Matthew S Fell.
.br
.BR wadext "(6), "
.BR wadflat "(6), "
.BR deutex "(6), "
.BR wadgc "(6), "
.BR tkwadcad "(6), "
.BR ppm "(5), "
.BR xwadtools (6)

.SH BUGS
None known.

.SH AUTHOR
\*. was written by Udo Munk (um@compuserve.com) and hacked by Andre Majorel
(amajorel@teaser.fr).


raw2ppm/raw2ppm.c
/************************************************************************/
/*      Copyright (C) 1998, 1999 by Udo Munk (um@compuserve.com)	*/
/*      Copyright (C) 1999 by Andre Majorel (amajorel@teaser.fr)        */
/*                                                                      */
/*      Permission to use, copy, modify, and distribute this software   */
/*      and its documentation for any purpose and without fee is        */
/*      hereby granted, provided that the above copyright notice        */
/*      appears in all copies and that both that copyright notice and   */
/*      this permission notice appear in supporting documentation.      */
/*      The author and contibutors make no representations about the    */
/*      suitability of this software for any purpose. It is provided    */
/*      "as is" without express or implied warranty.                    */
/************************************************************************/

/*
 *   Tool to convert raw flats data extracted from a WAD file into ppm files
 */

/*
   AYM 1999-10-06
     New options -c, -h, -I, -n, -o, -s, -v, -w.
     Can now process more than one file at a time.
     Name of output file is now implied.

   UM 1999-11-17
     Use Andre's my_getopt instead depending on libc getopt.
*/

#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>
#include <errno.h>

#include "mygetopt.h"

#define MAXLEN 256

static int process_file(const char *ifile, const char *ofilename, int width,
		int height, long pixels);
static void err(const char *fmt, ...);

extern const unsigned char doom_rgb[];
extern const unsigned char doom04_rgb[];
extern const unsigned char doom04t_rgb[];
extern const unsigned char doom05_rgb[];
extern const unsigned char *doom05t_rgb;
extern const unsigned char heretic_rgb[];
extern const unsigned char hexen_rgb[];
extern const unsigned char strife_rgb[];

static long offset     = 0;
static long pixels     = -1;
static int  width      = -1;
static int  height     = -1;
static int  verbose    = 0;
static int  interleave = 0;
static const unsigned char *pal = doom_rgb;

void usage(char *name)
{
	fprintf(stderr,
	    "Usage: %s [options] file ...\n",
	    name);
	fprintf (stderr,
"Options:\n"
"  -c game  Palette (doom|doom04|doom04t|doom05|doom05t|heretic|hexen|strife)\n"
"  -h n     Set image height to n pixels. Default 64.\n"
"  -I       Use Doom alpha interleaving scheme.\n"
"  -n n     Read only n bytes of file. Default up to EOF.\n"
"  -o file  Set output name. Default is input name with .ppm extension.\n"
"  -s n     Skip first n bytes of file. Default 0.\n"
"  -v       Verbose mode.\n"
"  -w n     Set image width to n pixels. Default 64.\n");
	exit(1);
}

int main(int argc, char **argv)
{
	char		*program;
	const char	*output_file_name = NULL;  /* -o */
	int		rc = 0;

	/* save program name for usage() */
	program = *argv;

	/* parse options */
	{
	int r;
	while ((r = my_getopt(argc, argv, "c:h:In:o:s:vw:")) != EOF) {
		if (r == '?')
			usage(program);
		else if (r == 'c') {
			if (! strcmp(my_optarg, "doom"))
				pal = doom_rgb;
			else if (! strcmp(my_optarg, "doom04"))
				pal = doom04_rgb;
			else if (! strcmp(my_optarg, "doom04t"))
				pal = doom04t_rgb;
			else if (! strcmp(my_optarg, "doom05"))
				pal = doom05_rgb;
			else if (! strcmp(my_optarg, "doom05t"))
				pal = doom05t_rgb;
			else if (! strcmp(my_optarg, "heretic"))
				pal = heretic_rgb;
			else if (! strcmp(my_optarg, "hexen"))
				pal = hexen_rgb;
			else if (! strcmp(my_optarg, "strife"))
				pal = strife_rgb;
			else
				usage(program);
		}
		else if (r == 'h') {
			char *p;
			height = strtol(my_optarg, &p, 0);
			if (*my_optarg == '\0' || *p != '\0' || height < 1) {
				err("-h wants a strictly positive integer");
				exit(1);
			}
		}
		else if (r == 'I') {
			interleave = 1;
		}
		else if (r == 'n') {
			char *p;
			pixels = strtol(my_optarg, &p, 0);
			if (*my_optarg == '\0' || *p != '\0' || pixels < 0) {
				err("-n wants a positive integer");
				exit(1);
			}
		}
		else if (r == 'o') {
			if (strlen(my_optarg) > MAXLEN) {
				err("-o: name too long");
				exit(1);
			}
			output_file_name = my_optarg;
		}
		else if (r == 's') {
			char *p;
			offset = strtol(my_optarg, &p, 0);
			if (*my_optarg == '\0' || *p != '\0' || offset < 0) {
				err("-s wants a positive integer");
				exit(1);
			}
		}
		else if (r == 'v') {
			verbose = 1;
		}
		else if (r == 'w') {
			char *p;
			width = strtol(my_optarg, &p, 0);
			if (*my_optarg == '\0' || *p != '\0' || width < 1) {
				err("-w wants a strictly positive integer");
				exit(1);
			}
		}
		else
			usage(program);  /* can't happen */
	}
	}

	/* must have at least one file name */
	if (argc - my_optind < 1)
		usage(program);

	/* -o not allowed if more than one file name */
	if (output_file_name != NULL && argc - my_optind > 1) {
		err("-o not allowed if more than one file name");
		exit(1);
	}

	for (; my_optind < argc; my_optind++) {
		const char *ifilename = argv[my_optind];
		char ofilename[MAXLEN + 1];

		if (strlen(ifilename) > MAXLEN) {
			err("%.256s: name too long", ifilename);
			err("%.256s: fatal error--skipping file", ifilename);
			rc = 2;
			continue;
		}
		/* if the -o option was given, use that. Else build output
		   file name by appending ".ppm" to input file name after
		   having stripped it of any ".lmp" extension */
		if (output_file_name != NULL) {
			strcpy(ofilename, output_file_name);
		}
		else {
			const char *const ext1 = ".lmp";
			const char *const ext2 = ".ppm";
			char *p;

			strcpy(ofilename, ifilename);
			p = ofilename + strlen(ofilename);
			if (strlen(ofilename) >= strlen(ext1)
				&& !strcmp(p - strlen(ext1), ext1)) {
				p -= strlen(ext1);
				*p = '\0';
			}
			if (strlen(ofilename) + strlen(ext2) > MAXLEN) {
				err("%.256s: name too long", ifilename);
				err("%.256s: fatal error--skipping file");
				rc = 2;
				continue;
			}
			strcpy(p, ext2);
		}
		if (verbose)
			printf("%.256s -> %.256s\n", ifilename, ofilename);
		if (process_file(ifilename, ofilename, width, height, pixels))
			rc = 2;
	}
	return rc;
}

/*
 *	process_file
 *	Read one input file and write one output file
 *	Return 0 on success, non-zero on failure
 */
static int process_file(const char *ifile, const char *ofilename, int width,
		int height, long pixels)
{
	FILE *ifp = NULL;
	FILE *ofp = NULL;
	int rc = 0;

	if ((ifp = fopen(ifile, "rb")) == NULL) {
		err("can't open \"%.256s\" for reading (%s)",
				ifile, strerror(errno));
		rc = 1;
		goto byebye;
	}

	/* number of bytes (pixels) to read from the file */
	if (pixels < 0) {
		int r = fseek(ifp, 0, SEEK_END);
		if (r) {
			err("%.256s: can't seek to EOF (%s)",
					ifile, strerror(errno));
			rc = 1;
			goto byebye;
		}
		pixels = ftell(ifp) - offset;
		rewind(ifp);
	}

	/* if neither width nor height is given, assume width is 64 */
	if (width < 0 && height < 0) {
		width = 64;
		height = pixels / width;
	}
	/* if only width is given, deduce height from width */
	else if (height < 0 && width > 0) {
		height = pixels / width;
	}
	/* and conversely */
	else if (width < 0 && height > 0) {
		width = pixels / height;
	}
	/* if both are specified, just do that (adjust pixels later) */
	else if (width > 0 && height > 0) {
		;
	}
	/* can't happen */
	else {
		err("report this bug to the maintainer ! (w=%d, h=%d)",
				width, height);
		exit(1);
	}

	if ((long) width * height < pixels) {
		err("%.256s: dropping %ld pixels",
			ifile, pixels - (long) width * height);
		pixels = (long) width * height;
	}
	else if ((long) width * height > pixels) {
		err("%.256s: not enough data for %d x %d = %ld pixels",
				width, height, (long) width * height);
		rc = 1;
		goto byebye;
	}

	/* write the PPM header */
	if ((ofp = fopen(ofilename, "wb")) == NULL) {
		err("can't open \"%.256s\" for writing (%s)",
				ofilename, strerror(errno));
		rc = 1;
		goto byebye;
	}
	fprintf(ofp, "P6\n");
	fprintf(ofp, "# CREATOR: raw2ppm\n");
	fprintf(ofp, "%d %d\n", width, height);
	fprintf(ofp, "255\n");

	/* honour the -s (skip) option */
	if (offset != 0) {
		int r = fseek(ifp, offset, SEEK_SET);
		if (r) {
			err("%.256s: can't seek to %08lXh (%s)",
					ifile, offset, strerror(errno));
			rc = 1;
			goto byebye;
		}
	}

	/* write the PPM pixel data */
	{
		/* Freaky Doom alpha interleaving algorithm. The input file
		   contains 4 frames, each as high as the final image but only
		   1/4th as wide. The first frame contains columns 0, 4, 8...
		   of the final image. The second frame contains columns 1,
		   5... The third frame contains columns 2, 6... The fourth
		   frame contains columns 3, 7... VGA framebuffer layout ?? */
		if (interleave) {
			unsigned char *buf = malloc(pixels);
			unsigned char *p   = buf;
			long n;
			if (buf == NULL) {
				err("%.256s: not enough memory", ifile);
				rc = 1;
				goto byebye;
			}
			for (n = 0; n < pixels; n++) {
				int c = getc(ifp);
				if (c == EOF) {
					err("%.256s: unexpected EOF", ifile);
					rc = 1;
					goto byebye;
				}
				*p = c;
				p += 4;
				if (p >= buf + pixels)  /* new frame */
					p -= pixels - 1;
			}
			for (n = 0; n < pixels; n++) {
				int o = 3 * buf[n];
				putc(pal[o],     ofp);
				putc(pal[o + 1], ofp);
				putc(pal[o + 2], ofp);
			}
			free(buf);
		}
		/* normal, non-interleaved data */
		else {
			long n;
			for (n = 0; n < pixels; n++) {
				int c = getc(ifp);
				if (c == EOF) {
					err("%.256s: unexpected EOF", ifile);
					rc = 1;
					goto byebye;
				}
				putc(pal[c * 3],     ofp);
				putc(pal[c * 3 + 1], ofp);
				putc(pal[c * 3 + 2], ofp);
			}
		}
	}

byebye:
	if (ifp != NULL)
		fclose(ifp);
	if (ofp != NULL)
		fclose(ofp);
	if (rc)
		err("%.256s: fatal error--skipping file");
	return rc;
}

/*
 *	err
 *	printf() a message to stderr, prefixed by "raw2ppm: "
 *	and followed by a newline.
 */
static void err(const char *fmt, ...)
{
	va_list args;
	fflush(stdout);
	fputs("raw2ppm: ", stderr);
	va_start(args, fmt);
	vfprintf(stderr, fmt, args);
	putc('\n', stderr);
	fflush(stderr);  /* not normally needed */
}
